splunk_add_on_ucc_framework/generators/file_generator.py

```diff
@@ -0,0 +1,106 @@
+import logging
+from abc import ABC, abstractmethod
+from os.path import realpath, sep
+from typing import Any, Dict, List, Union
+
+from jinja2 import Environment, FileSystemLoader, select_autoescape
+
+from splunk_add_on_ucc_framework.commands.modular_alert_builder.alert_actions_helper import (
+    write_file,
+)
+from splunk_add_on_ucc_framework.commands.rest_builder.global_config_builder_schema import (
+    GlobalConfigBuilderSchema,
+)
+from splunk_add_on_ucc_framework.global_config import GlobalConfig
+
+from . import file_const as fc
+
+__all__ = ["FileGenerator", "begin"]
+
+logger = logging.getLogger("ucc_gen")
+
+
+class FileGenerator(ABC):
+    __description__ = "DESCRIBE THE FILE THAT IS GENERATED"
+
+    def __init__(
+        self,
+        global_config: GlobalConfig,
+        input_dir: str,
+        output_dir: str,
+        **kwargs: Any,
+    ) -> None:
+        """
+        :param global_config: the GlobalConfig object that is validated and parsed
+        :param input_dir: the path to the source code of globalConfig.(json|yaml)
+        :param output_dir: the path to output/<addon_name> directory
+        :param ucc_dir: the path of source code of UCC framework
+        :param addon_name: the addon_name that is being generated
+
+        """
+        super().__init__()
+        self._global_config: GlobalConfig = global_config
+        self._input_dir = input_dir
+        self._output_dir = output_dir
+        self._template_dir = [(sep.join([kwargs["ucc_dir"], "templates"]))]
+        self._addon_name: str = kwargs["addon_name"]
+        self.writer = write_file
+        self._gc_schema: GlobalConfigBuilderSchema = (
+            GlobalConfigBuilderSchema(global_config)
+            if global_config is not None
+            else {}
+        )
+        self._set_attributes(**kwargs)
+
+    @abstractmethod
+    def _set_attributes(self, **kwargs: Any) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def generate(self) -> Dict[str, str]:
```

Comment:
I wonder if maybe it would be better to have a separate class for keeping track of the generated files? What do you think? This class could then contain the docs generation part and later, some additional logic for checking the content?

Reply:
This class is for generating files. For the files which are generated, we just need one object which we would later be used to compare (the implementation done under #1220) and output the results.

Reply:
Ok, so those 2 PR's will be connected. I understand.